astropy/table/table.py

```diff
@@ -1244,18 +1245,30 @@ def __setitem__(self, item, value):
             name = item
             # If this is a column-like object that could be added directly to table
             if isinstance(value, BaseColumn) or self._add_as_mixin_column(value):
+                # If we're setting a new column to a scalar, broadcast it.
```

Comment:
What about using an `empty_like` method in the final `else` block below and making this block handle only non-scalars.  I would definitely like to keep this code here as clean as possible.